### PR TITLE
localStorage push with some changes to the flow of the program.

### PR DESCRIPTION
**Pull request by norbit8**
- Added the localStorage functionality.
- Integrated the "Remove Courses" function with the localStorage.
- Integrated the "DarkMode" function with the localStorage.
- Integrated the "Reset" function with the localStorage.
-  I've made massive changes to the DarkMode, everytime the DarkMode button is pressed it sends a data to the storageLoader.js (content page), where it handles the request.
- Some changes to the manifest that doesn't really matter.

Basically we now have the "Remove Courses", "Dark Mode" and the "Reset" button ready to go.

### CODE REVIEW

**Comment by norbit8** [comment]
@all-contributors please add @norbit8  for infrastructure and code

File: content_scripts/courseRemover.js [hunk 1]
```diff
@@ -28,14 +28,16 @@
      * Adds buttons that removes courses from the courses pane
      */
     async function addCourseRemovingOption() {
-        courses_list = document.getElementsByClassName('type_course depth_3 contains_branch');
+        var courses_list = document.getElementsByClassName('type_course depth_3 contains_branch');
```
Comment (idan2468) [inline content_scripts/courseRemover.js:31]: Better use querySelector

File: content_scripts/courseRemover.js [hunk 2]
```diff
@@ -45,4 +47,5 @@
     }
 
     addCourseRemovingOption();
+    localStorage.setItem('Shit', 'Head');
```
Comment (idan2468) [inline content_scripts/courseRemover.js]: Remove redundant localStorage item

File: popup/popup.js [hunk 3]
```diff
@@ -26,11 +28,16 @@ function listenForClicks() {
                     return;
                 case "Reset":
                     // TODO: implement
+                    const tabs = await browser.tabs.query({
+                        currentWindow: true,
+                        active: true
+                    }).catch(onError);
+                    sendMessageToTabs(tabs, { "reset": "true" });
+                    alert("Please refresh your browser tab to apply.");
```
Comment (idan2468) [inline popup/popup.js:36]: Force refresh after resetting?

File: popup/popup.js [hunk 4]
```diff
@@ -49,28 +56,68 @@ function listenForClicks() {
     });
 }
 
-async function darkModeSwitch(){
-    if(window.darkMode){
-        await browser.tabs.removeCSS({file:"./../dark-mode/dark-mode.css"})
-    }else{
-        await browser.tabs.insertCSS({file:"./../dark-mode/dark-mode.css"})
```
Comment (idan2468) [inline popup/popup.js:56]: Not it's better to add the dark mode CSS class this way through a message from the content script to the background script or just do it in the background script?

File: popup/popup.js [hunk 5]
```diff
@@ -49,28 +56,68 @@ function listenForClicks() {
...
+        browser.tabs.sendMessage(
+            tab.id,
+            data
+        ).then(response => {
+            // alert(JSON.stringify(response));
+            if (response.DarkMode == "Off") {
+                window.darkMode = false;
+            }
+            else {
+                window.darkMode = true;
```
Comment (idan2468) [inline popup/popup.js]: window.darkMode = response.DarkMode !== "Off"

File: popup/popup.js [hunk 6]
```diff
@@ -49,28 +56,68 @@ function listenForClicks() {
...
+        const tabs = await browser.tabs.query({
+            currentWindow: true,
+            active: true
+        }).catch(onError);
+        sendMessageToTabs(tabs, { "DarkMode": "On" });
+    }
+    // window.darkMode = !window.darkMode
+}
+
+// to revisit... (I don't think we need this anymore)
```
Comment (idan2468) [inline popup/popup.js:98]: Yeah I think so too, great job of sending messages to and from background scripts

**Review by idan2468 — COMMENTED** [review]
You did a great job !!
I've written few comments for some questions I had.

File: popup/popup.js (hunk 5 above)
Comment (norbit8) [reply popup/popup.js]: I don't quite understand this comment, but just to clarify a bit.
The `window.darkMode` "static variable" is necessary for the `darkModeSwitch()` function below, and I don't have an access to the `response` in there. (the response is from the content script, namely `storageLoader.js`)

File: popup/popup.js (hunk 4 above)
Comment (norbit8) [reply popup/popup.js:56]: I've implemented it by sending a message from the background script (`popup.js`) to the content script (`storageLoader.js`).

File: content_scripts/courseRemover.js (hunk 1 above)
Comment (norbit8) [reply content_scripts/courseRemover.js:31]: Why? readability-wise?

File: content_scripts/courseRemover.js (hunk 2 above)
Comment (norbit8) [reply content_scripts/courseRemover.js]: Whoops, fixed 😅.

File: popup/popup.js (hunk 3 above)
Comment (norbit8) [reply popup/popup.js:36]: So I've actually tried to reload a fresh (default) localStorage save in real time, right after clicking reset, but for some reason it crashes the whole add-on.
P.S 
the whole thing is implemented on the `storageLoader.js` file.

**Review by norbit8 — COMMENTED** [review]
Answers

File: popup/popup.js (hunk 3 above)
Comment (idan2468) [reply popup/popup.js:36]: After resetting the localStorage have you tried using location.reload()?

File: content_scripts/courseRemover.js (hunk 1 above)
Comment (idan2468) [reply content_scripts/courseRemover.js:31]: As I know querySelector is more used nowadays because it's more flexible
but it doesn't have much importance...

File: popup/popup.js (hunk 4 above)
Comment (idan2468) [reply popup/popup.js:56]: So what the function "addDarkMode" is for?

File: popup/popup.js (hunk 3 above)
Comment (norbit8) [reply popup/popup.js:36]: Oh you want us to refresh the page for him? but If the user has something important that he didn't save (e.g. a forum post)
it will delete it.

File: popup/popup.js (hunk 5 above)
Comment (idan2468) [reply popup/popup.js]: Yeah I understand, just wrote a shortcut for the if-else statement

File: popup/popup.js (hunk 3 above)
Comment (idan2468) [reply popup/popup.js:36]: Yeah, you have got a point, so can we revert the page to the default?
Like removing the CSS file from the doc?

Comment (norbit8) [reply popup/popup.js:36]: Yes you can but as I said, I tried to implement it but failed.
I will try again though.
(for now if you refresh the page it will go back to normal)

File: popup/popup.js (hunk 4 above)
Comment (norbit8) [reply popup/popup.js:56]: This function is on the `storage.js` file and this function is the one which is actually injects the CSS to the page, I know it's a bit confusing but all of the functions in the `popup.js` are just "listeners" which call the content script (`storage.js`) to do the actual job

File: popup/popup.js (hunk 5 above)
Comment (norbit8) [reply popup/popup.js]: Oh I see I'll change it in a new commit, thanks!

File: popup/popup.js (hunk 4 above)
Comment (idan2468) [reply popup/popup.js:56]: > I've implemented it by sending a message from the background script (`popup.js`) to the content script (`storageLoader.js`).

Yeah, now I understand what do you mean.

**Review by idan2468 — APPROVED** [review]

File: manifest.json [hunk 7]
```diff
@@ -14,11 +14,14 @@
...
                 "bootstrap-5.0.0-beta3-dist/css/bootstrap.min.css"
             ],
-            "js": ["bootstrap-5.0.0-beta3-dist/js/bootstrap.min.js"
+            "js": ["/content_scripts/storageLoader.js", "bootstrap-5.0.0-beta3-dist/js/bootstrap.min.js"
             ],
             "run_at": "document_end"
         }
     ],
+    "background": {
+      "scripts": ["popup/index.html"]
```
Comment (idan2468) [inline manifest.json:23]: Why did you put under scripts, HTML file?

Comment (norbit8) [reply manifest.json:23]: My bad🤦🏻‍♂️🤦🏻‍♂️🤦🏻‍♂️, I'll delete it later. Good catch!
